cl/users/models.py

```diff
@@ -190,8 +259,211 @@ def generate_recap_email(user_profile: UserProfile, append: int = None) -> str:
+    :param message_id: The unique message id assigned by Amazon SES
+    :param bounceSubType: The subtype of the bounce, as determined
+     by Amazon SES
+    :param recipient_emails: a list of email addresses one per recipient
+     to whom the bounce notification pertains
+    :return: None
+    """
+    unexpected_events = ["Suppressed", "OnAccountSuppressionList"]
+    for email in recipient_emails:
+        if bounceSubType in unexpected_events:
+            # Handle unexpected bounceSubType events, log a warning
+            logging.warning(
+                f"Unexpected {bounceSubType} hard bounce for {email}"
+            )
+            # After log the event ban the email address
+            # Only ban email address if it hasn't been previously banned
+            EmailFlag.objects.get_or_create(
+                email_address=email,
+                flag_type="ban",
+                defaults={"reason": bounceSubType},
+            )
+        else:
+            # Any other hard bounce even only ban the email address
+            # Only ban email address if it hasn't been previously banned
+            EmailFlag.objects.get_or_create(
+                email_address=email,
+                flag_type="ban",
+                defaults={"reason": bounceSubType},
+            )
+
+
+def handle_soft_bounce(
+    message_id: str, bounceSubType: str, recipient_emails: List[str]
+) -> None:
+    """Handle a soft bounce notification received from SNS
+    :param message_id: The unique message id assigned by Amazon SES
+    :param bounceSubType: The subtype of the bounce, as determined
+     by Amazon SES
+    :param recipient_emails: a list of email addresses one per recipient
+     to whom the bounce notification pertains
+    :return: None
+    """
+
+    back_off_events = ["Undetermined", "General", "MailboxFull"]
+    small_only_events = ["MessageTooLarge", "AttachmentRejected"]
+
+    MAX_RETRY_COUNTER = 5
+    INITIAL_HOURS = 2
+
+    for email in recipient_emails:
+        if bounceSubType in back_off_events:
+            # Handle events that must trigger a backoff event
+            backoff_event = BackoffEvent.objects.filter(
+                email_address=email,
+            )
+            if not backoff_event.exists():
+                # Create a backoff event for an email address if not exists
+                # Initialize retry_counter and next_retry_date
+                next_retry_date = now() + timedelta(hours=INITIAL_HOURS)
+                BackoffEvent.objects.create(
+                    email_address=email,
+                    retry_counter=0,
+                    next_retry_date=next_retry_date,
+                )
+            else:
+                # If a previous backoff event exists
+                retry_counter = backoff_event[0].retry_counter
+                next_retry_date = backoff_event[0].next_retry_date
+
+                # Check if waiting period expired
+                if next_retry_date <= now():
```

Comment:
I think flipping this is a bit more clear, so it reads as:

> If the time now is after the retry date

Instead of:

> If the retry date is before now

Same meaning, but the first one makes more sense to me. So:

```suggestion
                if now() >= next_retry_date:
```

It *might* be even more clear to do:

```
if now() < next_retry_date:
    # It's not yet time to retry. 
    continue
```

Then you can outdent all that follows. 

I'm also curious, are we missing the `else` statement here? What if it's not yet time to send the email? Do we save it or something?

Reply:
oh yes it's more clear like you suggested `if now() >= next_retry_date:` 

I used this option instead of `if now() < next_retry_date:` because here we need to act when the waiting period is over,  so we don't have an `else`, here is a screenshot where I show this part of the code, the `else` is "Don't update the backoff event" so we are just ignoring it, as you could see we have a `Queue Email` function before entering to these `if `statements, so yes we are going to queue this email to send it later.
 I put a comment in the code where we are going to call this function `#TODO Queue email function`

![screen_1](https://user-images.githubusercontent.com/486004/159105942-70d122ec-56cc-4ad8-9a43-6c9fda14d3de.png)